tensorflow_probability/bayesian_neural_network/trainer/bayesian_neural_network.py

```diff
@@ -0,0 +1,344 @@
+  ax.set_title("weight stddevs")
+  ax.set_xlim([0, 1.])
+
+  fig.tight_layout()
+  canvas.print_figure(fname, format="png")
+  print("saved {}".format(fname))
+
+
+@tfgfile_wrapper
+def plot_heldout_prediction(input_vals, probs,
+                            fname, n=10, title=""):
+  """Save a PNG plot visualizing posterior uncertainty on heldout data.
+
+  Args:
+    input_vals: A `float`-like Numpy `array` of shape
+      `[num_heldout] + IMAGE_SHAPE`, containing heldout input images.
+    probs: A `float`-like Numpy array of shape `[num_monte_carlo,
+      num_heldout, num_classes]` containing Monte Carlo samples of
+      class probabilities for each heldout sample.
+    fname: Python `str` filename to save the plot to.
+    n: Python `int` number of datapoints to vizualize.
+    title: Python `str` title for the plot.
+  """
+  fig = figure.Figure(figsize=(9, 3*n))
+  canvas = backend_agg.FigureCanvasAgg(fig)
+  for i in range(n):
+    ax = fig.add_subplot(n, 3, 3*i + 1)
+    ax.imshow(input_vals[i, :].reshape(IMAGE_SHAPE[:-1]), interpolation="None")
+
+    ax = fig.add_subplot(n, 3, 3*i + 2)
+    for prob_sample in probs:
+      sns.barplot(np.arange(10), prob_sample[i, :], alpha=0.1, ax=ax)
+      ax.set_ylim([0, 1])
+    ax.set_title("posterior samples")
+
+    ax = fig.add_subplot(n, 3, 3*i + 3)
+    sns.barplot(np.arange(10), np.mean(probs[:, i, :], axis=0), ax=ax)
+    ax.set_ylim([0, 1])
+    ax.set_title("predictive probs")
+  fig.suptitle(title)
+  fig.tight_layout()
+
+  canvas.print_figure(fname, format="png")
+  print("saved {}".format(fname))
+
+
+def build_input_pipeline(mnist_data, batch_size, heldout_size):
+  """Build an Iterator switching between train and heldout data."""
+
+  # Build an iterator over training batches.
+  training_dataset = tf.data.Dataset.from_tensor_slices(
+      (mnist_data.train.images, np.int32(mnist_data.train.labels)))
+  training_batches = training_dataset.shuffle(
+      50000, reshuffle_each_iteration=True).repeat().batch(batch_size)
+  training_iterator = training_batches.make_one_shot_iterator()
+
+  # Build a iterator over the heldout set with batch_size=heldout_size,
+  # i.e., return the entire heldout set as a constant.
+  heldout_dataset = tf.data.Dataset.from_tensor_slices(
+      (mnist_data.validation.images,
+       np.int32(mnist_data.validation.labels)))
+  heldout_frozen = (heldout_dataset.take(heldout_size).
+                    repeat().batch(heldout_size))
+  heldout_iterator = heldout_frozen.make_one_shot_iterator()
+
+  # Combine these into a feedable iterator that can switch between training
+  # and validation inputs.
+  handle = tf.placeholder(tf.string, shape=[])
+  feedable_iterator = tf.data.Iterator.from_string_handle(
+      handle, training_batches.output_types, training_batches.output_shapes)
+  images, labels = feedable_iterator.get_next()
+
+  return images, labels, handle, training_iterator, heldout_iterator
+
+
+def build_fake_data(num_examples=10):
```

Comment:
Should we move this function to another python file that contains testing stuff?